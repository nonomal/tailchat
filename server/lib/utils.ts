import randomString from 'crypto-random-string';
import _ from 'lodash';
import urlRegex from 'url-regex';

/**
 * 返回电子邮箱的地址
 * @param email 电子邮箱
 * @returns 电子邮箱
 */
export function getEmailAddress(email: string) {
  return email.split('@')[0];
}

/**
 * 生成随机字符串
 * @param length 随机字符串长度
 */
export function generateRandomStr(length = 10): string {
  return randomString({ length });
}

export function generateRandomNumStr(length = 6) {
  return randomString({
    length,
    type: 'numeric',
  });
}

/**
 * 是否一个可用的字符串
 * 定义为有长度的字符串
 */
export function isValidStr(str: unknown): str is string {
  return typeof str == 'string' && str !== '';
}

/**
 * 判断是否是一个可用的url
 */
export function isValidUrl(str: unknown): str is string {
  return typeof str == 'string' && urlRegex({ exact: true }).test(str);
}

/**
 * 检测一个地址是否是一个合法的资源地址
 */
export function isValidStaticAssetsUrl(str: unknown): str is string {
  if (typeof str !== 'string') {
    return false;
  }

  const filename = _.last(str.split('/'));
  if (filename.indexOf('.') === -1) {
    return false;
  }

  return true;
}

/**
 * 休眠一定时间
 */
export function sleep(ms: number): Promise<void> {
  return new Promise((resolve) =>
    setTimeout(() => {
      resolve();
    }, ms)
  );
}

/**
 * 检查url地址是否匹配
 */
export function checkPathMatch(urlList: string[], url: string): boolean {
  const fuzzList = urlList.map((url) => url.replaceAll('/', '.'));
  const fuzzUrl = url.split('?')[0].replaceAll('/', '.');

  // 考虑到serviceName中间可能会有. 且注册的时候不可能把所有情况都列出来，因此进行模糊处理
  return fuzzList.includes(fuzzUrl);
}
