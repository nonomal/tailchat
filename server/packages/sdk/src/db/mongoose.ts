export { Types, isValidObjectId } from 'mongoose';
