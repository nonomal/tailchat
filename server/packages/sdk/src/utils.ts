const noConflictKey = '__';

/**
 * 因为微服务的名称中经常会有 `.` , 而 `.` 在一些场景(如lodash.set) 有特殊含义，因此增加一个工具用于解决这个问题
 */
export function encodeNoConflictServiceNameKey(key: string): string {
  return key.replaceAll('.', noConflictKey);
}

export function decodeNoConflictServiceNameKey(key: string): string {
  if (typeof key !== 'string') {
    return '';
  }

  return key.replaceAll(noConflictKey, '.');
}
