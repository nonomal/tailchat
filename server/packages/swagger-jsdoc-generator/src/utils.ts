import { MethodDeclaration, ParameterDeclaration, SyntaxKind } from 'ts-morph';

/**
 * 获取函数参数列表
 */
export function getMethodParameters(
  methodDeclaration: MethodDeclaration
): ParameterDeclaration[] {
  return methodDeclaration.getChildrenOfKind(SyntaxKind.Parameter);
}
