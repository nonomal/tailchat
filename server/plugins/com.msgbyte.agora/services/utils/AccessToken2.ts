import crypto from 'crypto';
import zlib from 'zlib';
const VERSION_LENGTH = 3;
const APP_ID_LENGTH = 32;

const getVersion = () => {
  return '007';
};

class Service {
  __type;
  __privileges;

  constructor(service_type) {
    this.__type = service_type;
    this.__privileges = {};
  }

  __pack_type() {
    const buf = new ByteBuf();
    buf.putUint16(this.__type);
    return buf.pack();
  }

  __pack_privileges() {
    const buf = new ByteBuf();
    buf.putTreeMapUInt32(this.__privileges);
    return buf.pack();
  }

  service_type() {
    return this.__type;
  }

  add_privilege(privilege, expire) {
    this.__privileges[privilege] = expire;
  }

  pack() {
    return Buffer.concat([this.__pack_type(), this.__pack_privileges()]);
  }

  unpack(buffer) {
    const bufReader = new ReadByteBuf(buffer);
    this.__privileges = bufReader.getTreeMapUInt32();
    return bufReader;
  }
}

const kRtcServiceType = 1;

class ServiceRtc extends Service {
  static kPrivilegeJoinChannel = 1;
  static kPrivilegePublishAudioStream = 2;
  static kPrivilegePublishVideoStream = 3;
  static kPrivilegePublishDataStream = 4;

  __channel_name;
  __uid;

  constructor(channel_name, uid) {
    super(kRtcServiceType);
    this.__channel_name = channel_name;
    this.__uid = uid === 0 ? '' : `${uid}`;
  }

  pack() {
    const buffer = new ByteBuf();
    buffer.putString(this.__channel_name).putString(this.__uid);
    return Buffer.concat([super.pack(), buffer.pack()]);
  }

  unpack(buffer) {
    const bufReader = super.unpack(buffer);
    this.__channel_name = bufReader.getString();
    this.__uid = bufReader.getString();
    return bufReader;
  }
}

const kRtmServiceType = 2;

class ServiceRtm extends Service {
  __user_id;

  constructor(user_id) {
    super(kRtmServiceType);
    this.__user_id = user_id || '';
  }

  pack() {
    const buffer = new ByteBuf();
    buffer.putString(this.__user_id);
    return Buffer.concat([super.pack(), buffer.pack()]);
  }

  unpack(buffer) {
    const bufReader = super.unpack(buffer);
    this.__user_id = bufReader.getString();
    return bufReader;
  }
}

(ServiceRtm as any).kPrivilegeLogin = 1;

const kFpaServiceType = 4;

class ServiceFpa extends Service {
  constructor() {
    super(kFpaServiceType);
  }

  pack() {
    return super.pack();
  }

  unpack(buffer) {
    const bufReader = super.unpack(buffer);
    return bufReader;
  }
}

(ServiceFpa as any).kPrivilegeLogin = 1;

const kChatServiceType = 5;

class ServiceChat extends Service {
  __user_id;

  constructor(user_id) {
    super(kChatServiceType);
    this.__user_id = user_id || '';
  }

  pack() {
    const buffer = new ByteBuf();
    buffer.putString(this.__user_id);
    return Buffer.concat([super.pack(), buffer.pack()]);
  }

  unpack(buffer) {
    const bufReader = super.unpack(buffer);
    this.__user_id = bufReader.getString();
    return bufReader;
  }
}

(ServiceChat as any).kPrivilegeUser = 1;
(ServiceChat as any).kPrivilegeApp = 2;

const kEducationServiceType = 7;

class ServiceEducation extends Service {
  __room_uuid;
  __user_uuid;
  __role;

  constructor(roomUuid, userUuid, role) {
    super(kEducationServiceType);
    this.__room_uuid = roomUuid || '';
    this.__user_uuid = userUuid || '';
    this.__role = role || -1;
  }

  pack() {
    const buffer = new ByteBuf();
    buffer.putString(this.__room_uuid);
    buffer.putString(this.__user_uuid);
    buffer.putInt16(this.__role);
    return Buffer.concat([super.pack(), buffer.pack()]);
  }

  unpack(buffer) {
    const bufReader = super.unpack(buffer);
    this.__room_uuid = bufReader.getString();
    this.__user_uuid = bufReader.getString();
    this.__role = bufReader.getInt16();
    return bufReader;
  }
}

(ServiceEducation as any).PRIVILEGE_ROOM_USER = 1;
(ServiceEducation as any).PRIVILEGE_USER = 2;
(ServiceEducation as any).PRIVILEGE_APP = 3;

class AccessToken2 {
  appId;
  appCertificate;
  issueTs;
  expire;
  salt;
  services;

  constructor(appId, appCertificate, issueTs, expire) {
    this.appId = appId;
    this.appCertificate = appCertificate;
    this.issueTs = issueTs || new Date().getTime() / 1000;
    this.expire = expire;
    // salt ranges in (1, 99999999)
    this.salt = Math.floor(Math.random() * 99999999) + 1;
    this.services = {};
  }

  __signing() {
    let signing = encodeHMac(
      new ByteBuf().putUint32(this.issueTs).pack(),
      this.appCertificate
    );
    signing = encodeHMac(new ByteBuf().putUint32(this.salt).pack(), signing);
    return signing;
  }

  __build_check() {
    const is_uuid = (data) => {
      if (data.length !== APP_ID_LENGTH) {
        return false;
      }
      const buf = Buffer.from(data, 'hex');
      return !!buf;
    };

    const { appId, appCertificate, services } = this;
    if (!is_uuid(appId) || !is_uuid(appCertificate)) {
      return false;
    }

    if (Object.keys(services).length === 0) {
      return false;
    }
    return true;
  }

  add_service(service) {
    this.services[service.service_type()] = service;
  }

  build() {
    if (!this.__build_check()) {
      return '';
    }

    const signing = this.__signing();
    let signing_info = new ByteBuf()
      .putString(this.appId)
      .putUint32(this.issueTs)
      .putUint32(this.expire)
      .putUint32(this.salt)
      .putUint16(Object.keys(this.services).length)
      .pack();
    Object.values(this.services).forEach((service: any) => {
      signing_info = Buffer.concat([signing_info, service.pack()]);
    });

    const signature = encodeHMac(signing, signing_info);
    const content = Buffer.concat([
      new ByteBuf().putString(signature).pack(),
      signing_info,
    ]);
    const compressed = zlib.deflateSync(content);
    return `${getVersion()}${Buffer.from(compressed).toString('base64')}`;
  }

  from_string(origin_token) {
    const origin_version = origin_token.substring(0, VERSION_LENGTH);
    if (origin_version !== getVersion()) {
      return false;
    }

    const origin_content = origin_token.substring(
      VERSION_LENGTH,
      origin_token.length
    );
    const buffer = zlib.inflateSync(new Buffer(origin_content, 'base64'));
    const bufferReader = new ReadByteBuf(buffer);

    const signature = bufferReader.getString();
    this.appId = bufferReader.getString();
    this.issueTs = bufferReader.getUint32();
    this.expire = bufferReader.getUint32();
    this.salt = bufferReader.getUint32();
    const service_count = bufferReader.getUint16();

    let remainBuf = bufferReader.pack();
    for (let i = 0; i < service_count; i++) {
      const bufferReaderService = new ReadByteBuf(remainBuf);
      const service_type = bufferReaderService.getUint16();
      const service = new (AccessToken2 as any).kServices[service_type]();
      remainBuf = service.unpack(bufferReaderService.pack()).pack();
      this.services[service_type] = service;
    }
  }
}

const encodeHMac: any = function (key, message) {
  return crypto.createHmac('sha256', key).update(message).digest();
};

const ByteBuf: any = function () {
  const that: any = {
    buffer: Buffer.alloc(1024),
    position: 0,
  };

  that.buffer.fill(0);

  that.pack = function () {
    const out = Buffer.alloc(that.position);
    that.buffer.copy(out, 0, 0, out.length);
    return out;
  };

  that.putUint16 = function (v) {
    that.buffer.writeUInt16LE(v, that.position);
    that.position += 2;
    return that;
  };

  that.putUint32 = function (v) {
    that.buffer.writeUInt32LE(v, that.position);
    that.position += 4;
    return that;
  };
  that.putInt32 = function (v) {
    that.buffer.writeInt32LE(v, that.position);
    that.position += 4;
    return that;
  };

  that.putInt16 = function (v) {
    that.buffer.writeInt16LE(v, that.position);
    that.position += 2;
    return that;
  };

  that.putBytes = function (bytes) {
    that.putUint16(bytes.length);
    bytes.copy(that.buffer, that.position);
    that.position += bytes.length;
    return that;
  };

  that.putString = function (str) {
    return that.putBytes(Buffer.from(str));
  };

  that.putTreeMap = function (map) {
    if (!map) {
      that.putUint16(0);
      return that;
    }

    that.putUint16(Object.keys(map).length);
    for (const key in map) {
      that.putUint16(key);
      that.putString(map[key]);
    }

    return that;
  };

  that.putTreeMapUInt32 = function (map) {
    if (!map) {
      that.putUint16(0);
      return that;
    }

    that.putUint16(Object.keys(map).length);
    for (const key in map) {
      that.putUint16(key);
      that.putUint32(map[key]);
    }

    return that;
  };

  return that;
};

const ReadByteBuf: any = function (bytes) {
  const that: any = {
    buffer: bytes,
    position: 0,
  };

  that.getUint16 = function () {
    const ret = that.buffer.readUInt16LE(that.position);
    that.position += 2;
    return ret;
  };

  that.getUint32 = function () {
    const ret = that.buffer.readUInt32LE(that.position);
    that.position += 4;
    return ret;
  };

  that.getInt16 = function () {
    const ret = that.buffer.readUInt16LE(that.position);
    that.position += 2;
    return ret;
  };

  that.getString = function () {
    const len = that.getUint16();

    const out = Buffer.alloc(len);
    that.buffer.copy(out, 0, that.position, that.position + len);
    that.position += len;
    return out;
  };

  that.getTreeMapUInt32 = function () {
    const map = {};
    const len = that.getUint16();
    for (let i = 0; i < len; i++) {
      const key = that.getUint16();
      const value = that.getUint32();
      map[key] = value;
    }
    return map;
  };

  that.pack = function () {
    const length = that.buffer.length;
    const out = Buffer.alloc(length);
    that.buffer.copy(out, 0, that.position, length);
    return out;
  };

  return that;
};

(AccessToken2 as any).kServices = {};
(AccessToken2 as any).kServices[kRtcServiceType] = ServiceRtc;
(AccessToken2 as any).kServices[kRtmServiceType] = ServiceRtm;
(AccessToken2 as any).kServices[kFpaServiceType] = ServiceFpa;
(AccessToken2 as any).kServices[kChatServiceType] = ServiceChat;
(AccessToken2 as any).kServices[kEducationServiceType] = ServiceEducation;

export {
  AccessToken2,
  ServiceRtc,
  ServiceRtm,
  ServiceFpa,
  ServiceChat,
  ServiceEducation,
  kRtcServiceType,
  kRtmServiceType,
  kFpaServiceType,
  kChatServiceType,
  kEducationServiceType,
};
