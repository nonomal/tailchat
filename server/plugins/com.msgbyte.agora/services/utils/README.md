Copy from https://github.com/AgoraIO/Tools/tree/master/DynamicKey/AgoraDynamicKey/nodejs
