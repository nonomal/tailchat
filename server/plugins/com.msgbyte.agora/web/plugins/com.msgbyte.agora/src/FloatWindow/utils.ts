import type {
  IAgoraRTCClient,
  ILocalAudioTrack,
  ILocalTrack,
  ILocalVideoTrack,
} from 'agora-rtc-react';

/**
 * 获取客户端本地轨道
 */
export function getClientLocalTrack(
  client: IAgoraRTCClient,
  trackMediaType: 'video'
): ILocalVideoTrack | null;
export function getClientLocalTrack(
  client: IAgoraRTCClient,
  trackMediaType: 'audio'
): ILocalAudioTrack | null;
export function getClientLocalTrack(
  client: IAgoraRTCClient,
  trackMediaType: string
): ILocalTrack | null {
  return (
    client.localTracks.find(
      (track) => track.trackMediaType === trackMediaType
    ) ?? null
  );
}
