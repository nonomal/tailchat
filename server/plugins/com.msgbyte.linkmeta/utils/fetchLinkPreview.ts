import { getLinkPreview } from 'link-preview-js';

/**
 * 请求管理
 */
const cacheRequestList: Record<string, Promise<any>> = {};

/**
 * 获取网页元数据信息
 * @param url 网址
 * @returns
 */
export async function fetchLinkPreview(url: string): Promise<any> {
  if (cacheRequestList[url]) {
    // 如果有正在请求的信息
    return Promise.resolve(cacheRequestList[url]);
  }

  const promise = getLinkPreview(url);
  cacheRequestList[url] = promise;

  return Promise.resolve(promise).finally(() => {
    setTimeout(() => {
      delete cacheRequestList[url];
    }, 2 * 1000); // 窗口期, 请求完毕后2s内依旧会复用原来的接口
  });

  // return promise;
}
