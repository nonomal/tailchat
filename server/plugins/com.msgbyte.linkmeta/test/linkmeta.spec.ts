import { createTestServiceBroker } from '../../../test/utils';
import LinkmetaService from '../services/linkmeta.service';
import { Types } from 'mongoose';
import _ from 'lodash';

describe('Test "plugin:com.msgbyte.linkinfo" service', () => {
  const { broker, service, insertTestData } =
    createTestServiceBroker<LinkmetaService>(LinkmetaService);

  describe('Test "plugin:com.msgbyte.linkmeta.fetch"', () => {
    test('normal', async () => {
      const url = 'https://www.baidu.com/?fortest';
      const meta: any = await broker.call('plugin:com.msgbyte.linkmeta.fetch', {
        url,
      });

      try {
        expect(meta).toHaveProperty('url', url);
        expect(meta).toHaveProperty('isCache', false);
        expect(meta).toHaveProperty('title');
        expect(meta).toHaveProperty('siteName');
        expect(meta).toHaveProperty('description');
        expect(meta).toHaveProperty('mediaType', 'website');
        expect(meta).toHaveProperty('contentType', 'text/html');
        expect(meta).toHaveProperty('images');
        expect(meta).toHaveProperty('videos');
        expect(meta).toHaveProperty('favicons');
        expect(meta).toMatchSnapshot();

        const metaWithCache: any = await broker.call(
          'plugin:com.msgbyte.linkmeta.fetch',
          {
            url,
          }
        );
        expect(metaWithCache).toHaveProperty('isCache', true);
      } finally {
        await service.adapter.model.deleteOne({
          url,
        });
      }
    });

    test('pure video', async () => {
      const url = 'https://www.w3schools.com/html/mov_bbb.mp4';
      const meta: any = await broker.call('plugin:com.msgbyte.linkmeta.fetch', {
        url,
      });

      expect(meta).toMatchSnapshot();
    });

    test('pure image', async () => {
      const url = 'https://www.w3schools.com/html/pic_trulli.jpg';
      const meta: any = await broker.call('plugin:com.msgbyte.linkmeta.fetch', {
        url,
      });

      expect(meta).toMatchSnapshot();
    });

    test('pure ogg', async () => {
      const url = 'https://www.w3schools.com/html/horse.ogg';
      const meta: any = await broker.call('plugin:com.msgbyte.linkmeta.fetch', {
        url,
      });

      expect(meta).toMatchSnapshot();
    });

    test('pure mp3', async () => {
      const url = 'https://www.w3schools.com/html/horse.mp3';
      const meta: any = await broker.call('plugin:com.msgbyte.linkmeta.fetch', {
        url,
      });

      expect(meta).toMatchSnapshot();
    });
  });
});
