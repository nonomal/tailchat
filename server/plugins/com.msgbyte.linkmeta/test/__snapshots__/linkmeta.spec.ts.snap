// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test "plugin:com.msgbyte.linkinfo" service Test "plugin:com.msgbyte.linkmeta.fetch" normal 1`] = `
Object {
  "contentType": "text/html",
  "description": undefined,
  "favicons": Array [
    "https://www.baidu.com/favicon.ico",
  ],
  "images": Array [],
  "isCache": false,
  "mediaType": "website",
  "siteName": undefined,
  "title": "",
  "url": "https://www.baidu.com/?fortest",
  "videos": Array [],
}
`;

exports[`Test "plugin:com.msgbyte.linkinfo" service Test "plugin:com.msgbyte.linkmeta.fetch" pure image 1`] = `
Object {
  "contentType": "image/jpeg",
  "favicons": Array [
    "https://www.w3schools.com/favicon.ico",
  ],
  "isCache": true,
  "mediaType": "image",
  "url": "https://www.w3schools.com/html/pic_trulli.jpg",
}
`;

exports[`Test "plugin:com.msgbyte.linkinfo" service Test "plugin:com.msgbyte.linkmeta.fetch" pure mp3 1`] = `
Object {
  "contentType": "audio/mpeg",
  "favicons": Array [
    "https://www.w3schools.com/favicon.ico",
  ],
  "isCache": true,
  "mediaType": "audio",
  "url": "https://www.w3schools.com/html/horse.mp3",
}
`;

exports[`Test "plugin:com.msgbyte.linkinfo" service Test "plugin:com.msgbyte.linkmeta.fetch" pure ogg 1`] = `
Object {
  "contentType": "video/ogg",
  "favicons": Array [
    "https://www.w3schools.com/favicon.ico",
  ],
  "isCache": true,
  "mediaType": "video",
  "url": "https://www.w3schools.com/html/horse.ogg",
}
`;

exports[`Test "plugin:com.msgbyte.linkinfo" service Test "plugin:com.msgbyte.linkmeta.fetch" pure video 1`] = `
Object {
  "contentType": "video/mp4",
  "favicons": Array [
    "https://www.w3schools.com/favicon.ico",
  ],
  "isCache": true,
  "mediaType": "video",
  "url": "https://www.w3schools.com/html/mov_bbb.mp4",
}
`;
